src/backend/distributed/deparser/qualify_statistics_stmt.c

```diff
@@ -159,8 +189,38 @@ QualifyAlterStatisticsOwnerStmt(Node *node)
 	if (list_length(nameList) == 1)
 	{
 		RangeVar *stat = makeRangeVarFromNameList(nameList);
-		Oid schemaOid = RangeVarGetCreationNamespace(stat);
+		Oid statsOid = get_statistics_object_oid(nameList, true);
+
+		if (!OidIsValid(statsOid))
+		{
+			return;
+		}
+
+		Oid schemaOid = GetStatsNamespaceOid(statsOid);
 		stat->schemaname = get_namespace_name(schemaOid);
 		stmt->object = (Node *) MakeNameListFromRangeVar(stat);
 	}
 }
+
+
+/*
+ * GetStatsNamespaceOid takes the id of a Statistics object and returns
+ * the id of the schema that the statistics object belongs to.
+ * Errors out if the stats object is not found.
+ */
+static Oid
+GetStatsNamespaceOid(Oid statsOid)
+{
+	HeapTuple heapTuple = SearchSysCache1(STATEXTOID, ObjectIdGetDatum(statsOid));
+	if (!HeapTupleIsValid(heapTuple))
+	{
+		ereport(ERROR, (errmsg("cache lookup failed for statistics "
+							   "object with oid %u", statsOid)));
+	}
+	FormData_pg_statistic_ext *statisticsForm =
+		(FormData_pg_statistic_ext *) GETSTRUCT(heapTuple);
+
+	ReleaseSysCache(heapTuple);
+
+	return statisticsForm->stxnamespace;
```

Comment:
The value you are returning might have been turned into garbade after you have released the `heapTuple` one line above. In general you can only read from the tuple as long as it hasn't been released.

```
	FormData_pg_statistic_ext *statisticsForm =
		(FormData_pg_statistic_ext *) GETSTRUCT(heapTuple);
	Oid result = statisticsForm->stxnamespace

	ReleaseSysCache(heapTuple);

	return result;
```